### DISCUSSION TITLE
Nested blocks, for extracting links

### DISCUSSION THREAD

**Discussion by llewellynthomas** [codex-team/editor.js#1879] [Q&A]
This looks like an amazing tool! We are investigating removing our TinyMCE editor from our CMS and are interested in using your project.

If our editors add links to resources on the web, we would like to be able to identify if these items still exist or if they are redirecting, so we can update/remove the links. Currently, with TinyMCE, this is really difficult.

I love the way the data is presented and would be amazing for us, but I was wondering if you can or plan to have nested blocks? Or if it's already supported. If an editor adds an inline link into a paragraph, is it possible to have this as a nested block?

Currently this is how your editor outputs a paragraph:
```json
   {
            "type" : "paragraph",
            "data" : {
                "text" : "There are dozens of <a href=\"https://github.com/editor-js\">ready-to-use Blocks</a> and the <a href=\"https://editorjs.io/creating-a-block-tool\">simple API</a> for creation any Block you need. For example, you can implement Blocks for Tweets, Instagram posts, surveys and polls, CTA-buttons and even games."
            }
        },
```

Possible future:
```json
{
  "type": "paragraph",
  "data": [
    {
      "text": "There are dozens of "
    },
    {
      "link": {
        "url": "https://github.com/editor-js",
        "text": "ready-to-use Blocks"
      }
    }
    {
      "text": " and the "
    },
    {
      "link": {
        "url": "https://editorjs.io/creating-a-block-tool",
        "text": "simple API"
      }
    },
    {
      "text": " for creation any Block you need. For example, you can implement Blocks for Tweets, Instagram posts, surveys and polls, CTA-buttons and even games."
    }
  ]
}
```

Looks like the paragraph tool is a plugin so we would be able to write and share our own version, but would this allow us to create our own custom data? And would this be something you are interested in supporting?

Thanks for your time.

And again this is an amazing project!

**Comment by tyazeez** [comment]
It will be nice to see this feature added to the editor. Feeling so excited about the possibilities of this editor. Open source rocks!!

**Comment by winston0410** [comment]
Same here.  Looking forward for nested block.

**Comment by webdeb** [comment]
I think my issue is related to this / could be fixed with the proposed solution:
When you press Shift + Enter while editing a paragraph, an `<br>` is added. 
I would expect just a new-line `\n`. or with the proposed structure:

```
type: "paragraph",
data: [
  { text: "Hello", styles: ["bold"] },
  { text: "World", styles: ["bold", "italic"] },
  { text: "!", styles: ["bold"] },
  { type: "line-break"},
  { text: "more text..."},
]
```

Yes, its a lot of data, but:

**The Quote from the base-concepts page:**

> What is clean data
... Editor.js returns **clean data** instead of **HTML-markup** ...

**Comment by hata6502** [comment]
https://github.com/codex-team/editor.js/issues/836

**Comment by horizon89** [comment]
If you just want to process links inside text paragraphs, you can do what I do:

Inside your JSON processing function, you can use a regex like so:

`text = text.replace(/<a href="(.*?)">(.*?)<\/a>/g, (m, $1, $2) => TextLinkProcess($1,$2) );`

Then, you can process the anchor text and the url:

`        function TextLinkProcess(url, anchorText) {
          console.log(url);
          console.log(anchorText);
          // do stuff then 'return' result
        }`

**Comment by alessiostalla** [comment]
Nested blocks in general are not limited to links. One could use nested blocks to embed all sorts of things into the text, without encoding it in (pseudo-)HTML which kind of defeats the purpose of editor.js. I think this issue can be subdivided into several concerns:
 * does the API in general allow this for a custom block? If not, what's missing?
 * does it make sense to change the paragraph block to adopt such a nested data layout (which would also be a nested component layout, as in, how they're organized in the code)?
 * what would the UI/UX for this look like?

Basically, I'm thinking about nesting inline blocks, for things like live-updated counters or code fragments automatically extracted from some repository or other kinds of live data that we want to insert in the middle of the text (as opposed to a separate paragraph).

**Comment by hata6502** [comment]
Please see my plugin for inline nesting: https://github.com/hata6502/editorjs-inline#readme

But... it may output a not-clean data.

**Comment by mandolyte** [comment]
We have some complex JSON that is deeply nested. While there are opportunities for it to be simplified, some things seem to be naturally modeled by nesting blocks. Examples include:
- Chapters within a book
- Sections within chapters
- Lists within lists

Based on the comments above, I don't think this is possible today. 

If this is true, is it on the roadmap? Are there other ways to model such "parent-child" relationships?

**Comment by xmontero** [comment]
Hi, late to the party! @mandolyte I'm also interested in the "deeply nested JSON to represent documents" that are "potentially editable by a human" (for example in editor.js).

I'm particularly thinking of 3 projects I have in hands that would benefeit of defining the web page in JSON and templateing it later into HTML while being able to edit the page with editor.js.

Those 3 projects I'm currently working in are:

1. A quotation site. A site where quotations are generated programatically but, before shown to a human, maybe an agent wants to "edit the quotation" manually to add, remove, hide, change something that the bot generated. The quotation has complex nested data. The agent would edit the final quotation in an editor like editor.js.

2. A slow-changing catalog site. A site where 10.000+ products are shown into the final user dumped from databases from multiple providers. But once the product is rendered into a page, maybe a marketer wants to edit the "product page" to fine tune the automatically-loaded descriptions, etc. The catalog contains nested data in the product definition, that is later on edited by a product-manager with editor.js.

3. A radio station website. A site with about 1 million news generated during the last few years. The news should be "immutable" once they are closed, so the "original" should not be edited. Once those 1 million pages are automatically rendered from the immutable originals, the most consumed news are submitted to conduct a manual review only for typos or adding a photo or a video for context. This edition could be done by editor.js, while the news rendering may contain complex nested data.

So, questions:

To all: Does editor.js now support for "nested blocks"?
To @mandolyte is there any way we may DM? I'm specially interested in undertanding more about your particular JSON structures in relation to editor.js but that could maybe be offtopic or non public.
